File: internal/retry/retry.go
package retry

import (
	"context"
	"fmt"
	"math"
	"math/rand/v2"
	"time"
)

type Config struct {
	MaxAttempts    int
	InitialDelay   time.Duration
	MaxDelay       time.Duration
	BackoffFactor  float64
	JitterFraction float64 // 0.0–1.0: fraction of delay to add as random jitter
}

func DefaultConfig() Config {
	return Config{
		MaxAttempts:    60,
		InitialDelay:   time.Second,
		MaxDelay:       30 * time.Second,
		BackoffFactor:  2.0,
		JitterFraction: 0.1,
	}
}

func (c Config) Validate() error {
	if c.MaxAttempts < 1 {
		return fmt.Errorf("max-attempts must be >= 1, got %d", c.MaxAttempts)
	}
	if c.InitialDelay <= 0 {
		return fmt.Errorf("initial-delay must be > 0, got %s", c.InitialDelay)
	}
	if c.MaxDelay < c.InitialDelay {
		return fmt.Errorf("max-delay (%s) must be >= initial-delay (%s)", c.MaxDelay, c.InitialDelay)
	}
	if c.BackoffFactor < 1.0 {
		return fmt.Errorf("backoff-factor must be >= 1.0, got %f", c.BackoffFactor)
	}
	if c.JitterFraction < 0 || c.JitterFraction > 1 {
		return fmt.Errorf("jitter-fraction must be in [0, 1], got %f", c.JitterFraction)
	}
	return nil
}

func Delay(cfg Config, attempt int) time.Duration {
	delay := float64(cfg.InitialDelay) * math.Pow(cfg.BackoffFactor, float64(attempt))
	if delay > float64(cfg.MaxDelay) {
		delay = float64(cfg.MaxDelay)
	}

	if cfg.JitterFraction > 0 {
		jitter := delay * cfg.JitterFraction * rand.Float64()
		delay += jitter
	}

	return time.Duration(delay)
}

type Result struct {
	Attempt int
	Err     error
}

func Do(ctx context.Context, cfg Config, fn func(ctx context.Context, attempt int) error) Result {
	for attempt := range cfg.MaxAttempts {
		err := fn(ctx, attempt)
		if err == nil {
			return Result{Attempt: attempt, Err: nil}
		}

		if attempt == cfg.MaxAttempts-1 {
			return Result{Attempt: attempt, Err: fmt.Errorf("all %d attempts failed, last error: %w", cfg.MaxAttempts, err)}
		}

		delay := Delay(cfg, attempt)
		select {
		case <-ctx.Done():
			return Result{Attempt: attempt, Err: fmt.Errorf("context cancelled after attempt %d: %w", attempt+1, ctx.Err())}
		case <-time.After(delay):
		}
	}

	return Result{Err: fmt.Errorf("max attempts reached")}
}

File: src/retry.rs
use std::time::{Duration, Instant};

pub struct Config {
    pub max_attempts: u32,
    pub initial_delay: Duration,
    pub max_delay: Duration,
    pub backoff_factor: f64,
    pub jitter_fraction: f64,
}

impl Config {
    pub fn validate(&self) -> Result<(), String> {
        if self.max_attempts < 1 {
            return Err(format!(
                "max-attempts must be >= 1, got {}",
                self.max_attempts
            ));
        }
        if self.initial_delay.is_zero() {
            return Err("initial-delay must be > 0".into());
        }
        if self.max_delay < self.initial_delay {
            return Err(format!(
                "max-delay ({:?}) must be >= initial-delay ({:?})",
                self.max_delay, self.initial_delay
            ));
        }
        if self.backoff_factor < 1.0 {
            return Err(format!(
                "backoff-factor must be >= 1.0, got {}",
                self.backoff_factor
            ));
        }
        if !(0.0..=1.0).contains(&self.jitter_fraction) {
            return Err(format!(
                "jitter-fraction must be in [0, 1], got {}",
                self.jitter_fraction
            ));
        }
        Ok(())
    }
}

pub fn delay(cfg: &Config, attempt: u32) -> Duration {
    let base = cfg.initial_delay.as_secs_f64() * cfg.backoff_factor.powi(attempt as i32);
    let capped = base.min(cfg.max_delay.as_secs_f64());
    let jitter = if cfg.jitter_fraction > 0.0 {
        capped * cfg.jitter_fraction * rand::random::<f64>()
    } else {
        0.0
    };
    Duration::from_secs_f64(capped + jitter)
}

pub struct RetryResult {
    pub attempt: u32,
    pub err: Option<String>,
}

pub fn do_retry<F>(cfg: &Config, deadline: Option<Instant>, mut f: F) -> RetryResult
where
    F: FnMut(u32) -> std::result::Result<(), String>,
{
    for attempt in 0..cfg.max_attempts {
        match f(attempt) {
            Ok(()) => return RetryResult { attempt, err: None },
            Err(e) => {
                if attempt == cfg.max_attempts - 1 {
                    return RetryResult {
                        attempt,
                        err: Some(format!(
                            "all {} attempts failed, last error: {}",
                            cfg.max_attempts, e
                        )),
                    };
                }
                let d = delay(cfg, attempt);
                if let Some(dl) = deadline {
                    if Instant::now() + d > dl {
                        return RetryResult {
                            attempt,
                            err: Some(format!("deadline exceeded after attempt {}", attempt + 1)),
                        };
                    }
                }
                std::thread::sleep(d);
            }
        }
    }
    RetryResult {
        attempt: 0,
        err: Some("max attempts reached".into()),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn test_config() -> Config {
        Config {
            max_attempts: 3,
            initial_delay: Duration::from_millis(10),
            max_delay: Duration::from_millis(100),
            backoff_factor: 2.0,
            jitter_fraction: 0.0,
        }
    }

    #[test]
    fn test_validate_ok() {
        assert!(test_config().validate().is_ok());
    }

    #[test]
    fn test_validate_max_attempts() {
        let mut cfg = test_config();
        cfg.max_attempts = 0;
        assert!(cfg.validate().is_err());
    }

    #[test]
    fn test_validate_initial_delay() {
        let mut cfg = test_config();
        cfg.initial_delay = Duration::ZERO;
        assert!(cfg.validate().is_err());
    }

    #[test]
    fn test_validate_max_delay() {
        let mut cfg = test_config();
        cfg.max_delay = Duration::from_millis(1);
        assert!(cfg.validate().is_err());
    }

    #[test]
    fn test_validate_backoff() {
        let mut cfg = test_config();
        cfg.backoff_factor = 0.5;
        assert!(cfg.validate().is_err());
    }

    #[test]
    fn test_validate_jitter() {
        let mut cfg = test_config();
        cfg.jitter_fraction = 1.5;
        assert!(cfg.validate().is_err());
    }

    #[test]
    fn test_delay_exponential() {
        let cfg = test_config();
        let d0 = delay(&cfg, 0);
        let d1 = delay(&cfg, 1);
        let d2 = delay(&cfg, 2);
        assert!(d1 > d0);
        assert!(d2 > d1);
    }

    #[test]
    fn test_delay_capped() {
        let cfg = test_config();
        let d = delay(&cfg, 100);
        assert!(d <= cfg.max_delay + Duration::from_millis(1));
    }

    #[test]
    fn test_do_success() {
        let cfg = test_config();
        let result = do_retry(&cfg, None, |_| Ok(()));
        assert!(result.err.is_none());
        assert_eq!(result.attempt, 0);
    }

    #[test]
    fn test_do_eventual_success() {
        let cfg = test_config();
        let result = do_retry(&cfg, None, |attempt| {
            if attempt < 2 {
                Err("not yet".into())
            } else {
                Ok(())
            }
        });
        assert!(result.err.is_none());
        assert_eq!(result.attempt, 2);
    }

    #[test]
    fn test_do_all_fail() {
        let cfg = test_config();
        let result = do_retry(&cfg, None, |_| Err("fail".into()));
        assert!(result.err.is_some());
        assert!(result.err.unwrap().contains("all 3 attempts failed"));
    }

    #[test]
    fn test_do_deadline() {
        let cfg = Config {
            max_attempts: 100,
            initial_delay: Duration::from_millis(50),
            max_delay: Duration::from_secs(1),
            backoff_factor: 1.0,
            jitter_fraction: 0.0,
        };
        let deadline = Instant::now() + Duration::from_millis(10);
        let result = do_retry(&cfg, Some(deadline), |_| Err("fail".into()));
        assert!(result.err.is_some());
        assert!(result.err.unwrap().contains("deadline"));
    }
}
